### PR TITLE
Cce analytic test gaugewave

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Adds the gauge wave CCE analytic worldtube solution.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spec

### CODE REVIEW

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp]: I'm confused. Is this Schwarzschild or gauge wave or some weird hybrid? Help :P Maybe you mean something like "A gauge wave on a Schwarzschild background in time-dependent..."? I'm confused on what exactly this solution is, sorry.

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp]: help, mass? Frequency?

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp:89]: Maybe state this in the doxygen comment? This is clear to me :)

File: tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_GaugeWave.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_GaugeWave.cpp:24]: for my own knowledge, why the low frequency choice?

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp]: I don't remember what we did in other solutions. This seems to border up on a `\warning` since using these functions without this knowledge could lead to a... frustrating... experience :P I'll leave it up to you to decide, but something to consider :)

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp:174]: I will kindly ask you change parentheses->brackets->braces as you get deeper and deeper. It's very difficult to line up which closing paren goes with which opening paren.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp]: I copied this note string from the most recently merged version, but I'm happy to promote them to warnings -- I certainly agree that overlooking this feature can make for a bad time.

File: tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_GaugeWave.cpp
Comment (moxcodes) [reply tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_GaugeWave.cpp:24]: just because it adds another lengthscale into the problem. if you have a high frequency, you have terms of different scales and make it harder to get a tight approximation in the test.

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (nilsvu) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp:10]: add `cmath`, `Utilities/ConstantExpressions.hpp`

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp:31]: Do you also apply a translation like Eqn (151) in this paper, or are you planning to? Perhaps add a note here in the docs, since you're referring to the paper

File: tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py]: Only a suggestion, entirely up to you: In math-heavy Python files it can be useful to `from numpy import pi, sin, cos, exp` so you don't have to sprinkle those `np.` everywhere.

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp
Comment (nilsvu) [inline tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp]: Perhaps take the tolerance as an argument to `test_spherical_metric` function, since the precision is problem-dependent? (If you have trouble with the `Args` parameter pack, you could take it as a `std::tuple<Args...>`) Up to you if you think this is worth the time.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp]: Okay, totally up to you! Just figured I'd bring it up :) I'm fine either way

File: tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_GaugeWave.cpp
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_GaugeWave.cpp:24]: Ah okay,thank you!

File: tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py]: but namespacing.... I find this is an interesting suggestion since it goes against the advice given in the C++ community :) Is the advice of "don't do `using namespace...`" not a thing in Python?

Comment (moxcodes) [reply tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py]: My understanding from a mild exposure to python is that when you do `from blah import foo` stuff, it's actually regarded as 'good enough' for the namespace philosophy, because it is always explicitly in the file what you are using, and a lot of editors (thankfully emacs too...) have good packages for working out the set of objects and functions that are brought in from such calls.
My understanding is that it is considered less accepted practice to do the (nonetheless fairly common) `from numpy import *` because it's impossible to work out the full set of garbage that brings into the local scope.

Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py]: ah yea, that makes sense. doing `using std::max` is a lot less bad than bringing in the whole namespace. Thanks for the explanation! :)

Comment (nilsvu) [reply tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py]: Exactly, in fact most (all?) linters will shout loudly at you if you try to `from x import *`

**Comment by moxcodes** [comment]
fixups posted, though I'll wait until @keefemitman gets a chance to submit his review before I squash.
Cheers!

**Comment by keefemitman** [comment]
> fixups posted, though I'll wait until @keefemitman gets a chance to submit his review before I squash.
> Cheers!

Sorry! I'm planning to review first thing tomorrow morning.

**Comment by moxcodes** [comment]
No rush, I just don't want to make the changes appear too confusing :)

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (keefemitman) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp]: I don't think there should be a frequency factor on this term

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (keefemitman) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp:101]: this is just for my own clarification, but we scale out the factor of sine, correct?

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (keefemitman) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp:125]: I would just reference the equation this comes from, i.e., Eq. (149)

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (keefemitman) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp:101]: Ah I see this is explained in the .hpp file, never mind!

File: tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py
Comment (keefemitman) [inline tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.py]: Again, I don't think this factor of f should be here.

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (keefemitman) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp:82]: this is just for my own clarification, but since this is a tnsr::aa object I assume that it's forced to be symmetric?

**Review by keefemitman — CHANGES_REQUESTED** [review]
The main issue I have is that I believe there is an incorrect factor of frequency in du_du_F(u) in both the C++ and python code. I also have some questions just for my own clarification, but otherwise everything looks good!

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp:82]: Yeah, that's right -- in fact, in spectre the symmetric tensors only have (N^2 + N)/2 storage slots in memory, so the `get<0, 1>(tensor)` points to literally the same data as `get<1, 0>(tensor)`.

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp:125]: I'll add the reference, but I think it's probably best to keep the explicit form here as well, just because there are some alterations in the precise form to e.g. minimize division operations to improve anticipated efficiency, and it's easiest for verification if the implementation is identical to the tex to compare with.

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp]: good catch! I agree.

**Comment by moxcodes** [comment]
Posted a fixup commit to address @keefemitman 's comments -- let me know if the changes are satisfactory.
Cheers!

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp
Comment (nilsvu) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp]: This is related to our discussion here https://github.com/sxs-collaboration/spectre/pull/2453#discussion_r485714390. Without a test that checks the solution actually solves the system it's very easy to introduce a bug like this in both the C++ and Python implementation. @moxcodes you mentioned you had such tests planned. I'd be interested to hear what you have in mind, perhaps in one of the next Monday meetings? For reference, here's how such tests are implemented for elliptic systems: https://github.com/sxs-collaboration/spectre/blob/develop/tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp

Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.cpp]: I'm happy to discuss in one of the meetings, but to summarize the basic idea -- the tests here are analytic only in the sense that we know ahead of time the values for the worldtube data and the resulting news, but not detailed values in the CCE domain (due to the complexities of the CCE null-geodesic coordinate system).
So, the way I plan to test these is via a complete test executable that performs the full evolution, and produces H5 outputs for the deviation from expectation. Those will be quick evolutions, but probably not quick enough to justify putting them in the unit tests.

File: src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/AnalyticSolutions/GaugeWave.hpp]: space needed between period and `The gauge per...`

**Review by nilsdeppe — COMMENTED** [review]
I'm happy with this. One tiny typo. You can squash everything :)

**Comment by moxcodes** [comment]
@keefemitman @nilsleiffischer May I squash?

**Comment by keefemitman** [comment]
Yes!

Kind regards,
Keefe Mitman
—
Keefe E. A. Mitman
California Institute of Technology
PhD Candidate in Physics
Pronouns in use: he, him, his

On Nov 10, 2020, at 12:36 PM, Jordan Moxon <notifications@github.com> wrote:

﻿

@keefemitman<https://github.com/keefemitman> @nilsleiffischer<https://github.com/nilsleiffischer> May I squash?

**Comment by nilsvu** [comment]
yes please squash!

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Comment by nilsdeppe** [comment]
formatting check failed

**Review by nilsvu — APPROVED** [review]
Please don't wait for my re-approval when the formatting check is fixed, rebased etc

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
